### PR TITLE
Biosys V3

### PR DESCRIPTION
**Pull request by serge-gaia**
This is a massive MR. Pretty much unreadable.
Here is the summary:
- Removed old legacy code
- Simplified model (all in `main.models`): Project, Site, Dataset, GenericRecord, Observation, SpeciesObservation.
- Django Rest Framework on top.
- Port to python3: meant to run on python3 (still py2 backward compatible for the moment)
- Good test coverage of the API.

TODO
- Views (admin/upload/download) to be ported to js+api only.
- More elaborate species management.
- See TODOs in the code.

Warning: **needs a database reset** some initial data (just for testing) can be loaded with the  `load_test_fixtures` command.

### CODE REVIEW

File: biosys/apps/main/admin.py [hunk 1]
```diff
@@ -1,151 +1,45 @@
...
-
 
 @admin.register(Site)
 class SiteAdmin(MainAppAdmin, GeoModelAdmin):
     change_form_template = 'main/site_change_form.html'
+    fields = ('project', 'parent_site',  'code', 'name', 'geometry', 'comments', 'attributes')
     list_display = [
-        'site_code', 'project', 'site_ID', 'site_name', 'date_established', 'established_by',
-        'parent_site']
+        'code', 'project', 'name', 'parent_site']
```
Comment (tony-gaia) [inline biosys/apps/main/admin.py]: Move to previous line for consistency

File: biosys/apps/main/api/serializers.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            site_fk = schema.get_fk_for_model('Site')
+            if site_fk:
+                model_field = site_fk.model_field
+                site_value = data.get(site_fk.data_field)
+                kwargs = {
+                    "project": self.dataset.project,
+                    model_field: site_value
+                }
+                site = Site.objects.filter(**kwargs).first()
+                if site is None:
```
Comment (tony-gaia) [inline biosys/apps/main/api/serializers.py]: could combine the above two statements to:

`if not Site.objects.filter(**kwargs).exists():`

File: biosys/apps/main/api/serializers.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                    msg = "Could not find the site '{} in: {}':".format(site_value, data)
+                    raise ValidationError(msg)
+
+    def set_context(self, serializer_field):
```
Comment (tony-gaia) [inline biosys/apps/main/api/serializers.py:71]: Maybe this should be called `set_dataset` (or does is have to have this name?)

File: biosys/apps/main/admin.py (hunk 1 above)
Comment (serge-gaia) [reply biosys/apps/main/admin.py]: yep

File: biosys/apps/main/api/serializers.py (hunk 2 above)
Comment (serge-gaia) [reply biosys/apps/main/api/serializers.py]: Always forget this one.

File: biosys/apps/main/api/serializers.py (hunk 3 above)
Comment (serge-gaia) [reply biosys/apps/main/api/serializers.py:71]: Yes it's a convention for  a Validator in DRF:
[http://www.django-rest-framework.org/api-guide/validators/#using-set_context](url)

File: biosys/apps/main/api/urls.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from rest_framework import routers
+
+from main.api import views as main_views
+
+
+router = routers.DefaultRouter()
+router.register(r'projects?', main_views.ProjectViewSet, 'project')
+router.register(r'sites?', main_views.SiteViewSet, 'site')
+router.register(r'datasets?', main_views.DatasetViewSet, 'dataset')
+router.register(r'genericRecords?', main_views.GenericRecordViewSet, 'genericRecord')
```
Comment (tony-gaia) [inline biosys/apps/main/api/urls.py]: How come using camel case rather than snake case?

File: biosys/apps/main/models.py [hunk 5]
```diff
@@ -68,60 +70,141 @@ def schema_model(self):
...
+        Check that the user is a custodian of the project pk passed in the POST data
+        :param request:
+        :return:
+        """
+        result = False
+        if is_admin(request.user):
+            result = True
+        else:
+            try:
+                project = Project.objects.get(pk=request.data['project'])
```
Comment (tony-gaia) [inline biosys/apps/main/models.py:183]: Serge is using `objects.get`?!? Has the world gone topsy turvey?

I usually follow it with a Project.DoesNotExist exception (or similar) for clarity.

File: biosys/apps/main/models.py [hunk 6]
```diff
@@ -180,458 +322,185 @@ class Observation(AbstractObservationRecord):
...
-
-
-def _calculate_site_ID():  # @NoSelf
-    if Site.objects.count() == 0:
-        return DEFAULT_SITE_ID
-    else:
-        return Site.objects.aggregate(Max('site_ID'))['site_ID__max'] + 1
-
 
+@python_2_unicode_compatible
```
Comment (tony-gaia) [inline biosys/apps/main/models.py:423]: Do you still need these decorators if the project is to run under Python 3? or is this a way of enabling it to run under Python 2 as well?

Comment (serge-gaia) [reply biosys/apps/main/models.py:423]: Yep currently the code py2/py3 compatible. When I'm sure that OIM can run python3 I will probably remove all the py2 stuff (like the `six` module I use from time to time)

File: biosys/apps/main/models.py (hunk 5 above)
Comment (serge-gaia) [reply biosys/apps/main/models.py:183]: :-) I don't know what kind of evil got me.
Why I don't specify the exception is that I want also to catch the error if 'project' in not in the `request.data`. It's laziness.

File: biosys/apps/main/api/urls.py (hunk 4 above)
Comment (serge-gaia) [reply biosys/apps/main/api/urls.py]: I don't know. I changed my mind a couple of times. I wanted to match the model name, but now I realize that snake case is probably better for URLs. I will change that.

File: biosys/apps/main/tests/api/test_observation.py [hunk 7]
```diff
@@ -0,0 +1,798 @@
...
+        self.assertEquals(ds.record_queryset.first().site, expected_site)
+
+    def test_update_site(self):
+        ds = self.ds_1
+        record = ds.record_queryset.filter(site=self.site_1).first()
+        self.assertIsNotNone(record)
+        site = Site.objects.filter(name="Site1").first()
+        # need to test if the site belongs to the dataset project or the update won't happen
+        self.assertIsNotNone(site)
+        self.assertTrue(site.project == record.dataset.project)
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_observation.py]: `assertEquals`?

File: biosys/apps/main/tests/api/test_site.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            for url in urls:
+                # site code must be unique
+                for site in data:
+                    site['code'] += '1'
+                self.assertEqual(
+                    client.post(url, data, format='json').status_code,
+                    status.HTTP_201_CREATED
+                )
+
+    def test_bulk_create_sneaky(self):
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_site.py:164]: nice name ;)

File: biosys/apps/main/tests/api/test_species_observation.py [hunk 9]
```diff
@@ -0,0 +1,908 @@
...
+            }
+            client = self.custodian_1_client
+            count = ds.record_queryset.count()
+            self.assertEqual(
+                client.post(url_post, data, format='json').status_code,
+                status.HTTP_201_CREATED
+            )
+            self.assertEquals(ds.record_queryset.count(), count + 1)
+
+        invalid_values = [None, '', 'not a date']
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_species_observation.py]: shouldn't the last be 'not a geom'?

File: biosys/apps/main/tests/api/test_species_observation.py [hunk 10]
```diff
@@ -0,0 +1,908 @@
...
+            status.HTTP_201_CREATED
+        )
+        self.assertEquals(ds.record_queryset.count(), 1)
+        self.assertEquals(ds.record_queryset.first().species_name, expected_name)
+
+    def test_update(self):
+        """
+        Test that the date and geometry are extracted from the data
+        and saved in DB
+        :return:
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_species_observation.py:765]: I think you need to update this comment

File: biosys/apps/main/tests/api/test_species_observation.py [hunk 11]
```diff
@@ -0,0 +1,908 @@
...
+        self.readonly_user.save()
+        self.readonly_client = APIClient()
+        self.assertTrue(self.readonly_client.login(username=self.readonly_user.username, password=password))
+
+        self.anonymous_client = APIClient()
+
+    def test_create(self):
+        """
+        Test that the date and geometry are extracted from the data
+        and saved in DB
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_species_observation.py]: Update description comment

File: biosys/apps/main/tests/api/test_species_observation.py [hunk 12]
```diff
@@ -0,0 +1,908 @@
...
+                status.HTTP_201_CREATED
+            )
+            self.assertEquals(ds.record_queryset.count(), 1)
+            self.assertEquals(ds.record_queryset.first().name_id, name_id)
+
+    def test_update(self):
+        """
+        Test that the date and geometry are extracted from the data
+        and saved in DB
+        :return:
```
Comment (tony-gaia) [inline biosys/apps/main/tests/api/test_species_observation.py:885]: and again

File: biosys/apps/main/utils_species.py [hunk 13]
```diff
@@ -64,53 +65,77 @@ def __init__(self, herbie_name):
...
+        NOTE: limiting the number of properties speed-up the request.
+        """
+        raise NotImplementedError('`get_all_species(properties` must be implemented.')
+
+
+class HerbieFacade(SpeciesFacade):
+    @staticmethod
+    def _add_attributes_filter_to_params(properties, params=None):
+        """
+        :param properties: al sequence of Property
```
Comment (tony-gaia) [inline biosys/apps/main/utils_species.py]: misspelling?

File: biosys/apps/main/utils_species.py [hunk 14]
```diff
@@ -64,53 +65,77 @@ def __init__(self, herbie_name):
...
+        :param params: an initial parameter dictionary
+        :return:
+        """
+        if properties:
+            if params is None:
+                params = {}
+                # WFS spec: {'propertyName': (p1,p2,..)}
+            params['propertyName'] = "({list})".format(
```
Comment (tony-gaia) [inline biosys/apps/main/utils_species.py]: best not to use reserved keywords for string formatting operations

**Review by tony-gaia — CHANGES_REQUESTED** [review]
Just some comments on the last bunch of files that github can't open because there were too many other changes to display:

settings.py:
Can you remove the tastypi settings?

feedback.html:
Tabbing is a bit out of whack

login.html:
Tabbing is a bit out of whack

wsgi.py:
put import confy and import os above the other import statements.

File: biosys/apps/main/tests/api/test_species_observation.py (hunk 12 above)
Comment (serge-gaia) [reply biosys/apps/main/tests/api/test_species_observation.py:885]: fixed

File: biosys/apps/main/tests/api/test_species_observation.py (hunk 11 above)
Comment (serge-gaia) [reply biosys/apps/main/tests/api/test_species_observation.py]: fixed

File: biosys/apps/main/tests/api/test_species_observation.py (hunk 10 above)
Comment (serge-gaia) [reply biosys/apps/main/tests/api/test_species_observation.py:765]: fixed

File: biosys/apps/main/tests/api/test_species_observation.py (hunk 9 above)
Comment (serge-gaia) [reply biosys/apps/main/tests/api/test_species_observation.py]: fixed

File: biosys/apps/main/tests/api/test_observation.py (hunk 7 above)
Comment (serge-gaia) [reply biosys/apps/main/tests/api/test_observation.py]: fixed

File: biosys/apps/main/utils_species.py (hunk 13 above)
Comment (serge-gaia) [reply biosys/apps/main/utils_species.py]: yep

File: biosys/apps/main/utils_species.py (hunk 14 above)
Comment (serge-gaia) [reply biosys/apps/main/utils_species.py]: I don't really see why but you're the boss ;-)

Comment (serge-gaia) [reply biosys/apps/main/utils_species.py]: fixed

**Pull request merged**